### PR TITLE
Enhance/Refactor Oeq Properties

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes [#7162](https://github.com/apache/arrow-datafusion/issues/7162).

## Rationale for this change
Ordering equivalence now considers constants during normalization, with this change we can use information after filter (when result is constant after filtering) to produce better plans.

While trying to close [#7162](https://github.com/apache/arrow-datafusion/issues/7162). I realized that, having `OrderingEquivalence` and `Equivalence` as different instantiations of same generic is a bit constraining. I took this opportunity to refactor `OrderingEquivalence` implementation (Moving functions to struct methods, and keeping track of constants so that we can use this information during normalization to produce better plans.).

## Are these changes tested?
Yes new tests are added to show that constants are ignored during Sort analysis.

### CODE REVIEW

File: datafusion/physical-expr/src/equivalence.rs [hunk 1]
```diff
@@ -131,17 +210,115 @@ impl<T: Eq + Clone + Hash> EquivalenceProperties<T> {
 /// where both `a ASC` and `b DESC` can describe the table ordering. With
 /// `OrderingEquivalenceProperties`, we can keep track of these equivalences
 /// and treat `a ASC` and `b DESC` as the same ordering requirement.
-pub type OrderingEquivalenceProperties = EquivalenceProperties<LexOrdering>;
+#[derive(Debug, Clone)]
+pub struct OrderingEquivalenceProperties {
+    oeq_class: Option<OrderingEquivalentClass>,
+    /// Keeps track of expressions that have constant value.
+    constants: Vec<Arc<dyn PhysicalExpr>>,
```
Comment (berkaysynnada) [inline datafusion/physical-expr/src/equivalence.rs:217]: Maybe `constants: HashSet<Arc<dyn PhysicalExpr>>` would be more practical.

File: datafusion/physical-expr/src/equivalence.rs [hunk 2]
```diff
@@ -106,14 +110,89 @@ impl<T: Eq + Clone + Hash> EquivalenceProperties<T> {
...
+                        }
+                        None
+                    });
+                Ok(if let Some(normalized_form) = normalized_form {
+                    Transformed::Yes(normalized_form)
+                } else {
+                    Transformed::No(expr)
+                })
+            })
+            .unwrap_or(expr)
```
Comment (metesynnada) [inline datafusion/physical-expr/src/equivalence.rs:142]: Why did we silence the error coming from here?

File: datafusion/physical-expr/src/equivalence.rs [hunk 3]
```diff
@@ -131,17 +210,115 @@ impl<T: Eq + Clone + Hash> EquivalenceProperties<T> {
...
+                self.constants.push(constant);
+            }
+        }
+    }
+
+    pub fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    pub fn normalize_sort_requirements(
```
Comment (metesynnada) [inline datafusion/physical-expr/src/equivalence.rs:274]: Docstring

File: datafusion/core/src/physical_plan/joins/utils.rs [hunk 4]
```diff
@@ -411,20 +367,28 @@ pub fn combine_join_ordering_equivalence_properties(
...
-                let right_oeq_classes =
-                    get_updated_right_ordering_equivalence_properties(
-                        join_type,
-                        right_oeq_properties.classes(),
-                        left_columns_len,
-                        &join_eq_properties,
-                    )?;
+            if let (
+                Some(JoinSide::Left),
+                Some(_right_ordering),
```
Comment (metesynnada) [inline datafusion/core/src/physical_plan/joins/utils.rs:374]: `Some(_)`

File: datafusion/core/src/physical_plan/joins/utils.rs [hunk 5]
```diff
@@ -433,32 +397,43 @@ pub fn combine_join_ordering_equivalence_properties(
...
+            new_properties.extend(updated_right_oeq);
             // In this special case, left side ordering can be prefixed with right side ordering.
-            if probe_side == Some(JoinSide::Right)
-                && left.output_ordering().is_some()
-                && *join_type == JoinType::Inner
-            {
-                let left_oeq_classes = left_oeq_properties.classes();
+            if let (
+                Some(JoinSide::Right),
+                Some(_left_ordering),
```
Comment (metesynnada) [inline datafusion/core/src/physical_plan/joins/utils.rs:424]: `Some(_)`

File: datafusion/physical-expr/src/analysis.rs [hunk 6]
```diff
@@ -231,7 +231,8 @@ fn calculate_selectivity(
                 1.0,
                 |acc, (i, ExprBoundaries { interval, .. })| {
                     let temp =
-                        cardinality_ratio(&initial_boundaries[i].interval, interval)?;
+                        cardinality_ratio(&initial_boundaries[i].interval, interval)
```
Comment (metesynnada) [inline datafusion/physical-expr/src/analysis.rs:234]: This error is being silenced. Is it because of float division?

**Review by metesynnada — APPROVED** [review]
Minor comments.

File: datafusion/physical-expr/src/equivalence.rs (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/equivalence.rs:217]: Unfortunately when using HashSet I cannot insert inside entries

File: datafusion/physical-expr/src/equivalence.rs (hunk 2 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/equivalence.rs:142]: If we can't normalize, we return original expression.

File: datafusion/physical-expr/src/analysis.rs (hunk 6 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/analysis.rs:234]: cardinality can be between 0.0 and 1.0. If we get an error, we just use 1.0 for cardinality estimate (safest estimate).

**Comment by mustafasrepo** [comment]
Merged upstream in [PR7566](https://github.com/apache/arrow-datafusion/pull/7566)

**Pull request closed without merging**